### ISSUE TITLE
X2C numerical instability

### ISSUE THREAD

**Issue by penglinq** [pyscf/pyscf#2209]
The X matrix has extremely large elements (>1e20) for some systems with PBC, likely because the A and B matrices in B = XA are rank deficient and thus X cannot be solved with numpy.linalg.solve as below: https://github.com/pyscf/pyscf/blob/940e4ac16f02eeef3fc944eae190d7f0609a60e7/pyscf/x2c/x2c.py#L880 A and B refer to cl and cs matrices in the code. This then leads to very negative energy (>100 times more negative than the expected energy)

Here is an example to reproduce the problem, a single Dy atom in a PBC box of 15A x 15A x 15A. 
```
from pyscf.pbc import gto, scf
import numpy as np

cell = gto.M(
    atom = '''Dy 0 0 0''',
    basis = 'sarc-dkh',
    a = np.eye(3)*15,
    unit = 'A',
    charge = 0,
    verbose = 6,
    )
mf = scf.GHF(cell, [0,0,0]).density_fit().x2c()
mf.kernel()
````
numpy.max(numpy.abs(x)) diverged (>1e20) and the rank of cl and cs were 2 less than the full rank in my test. I'm not sure why but numpy.linalg.solve did not raise a numpy.linalg.LinAlgError. 

My first question is, is rank-deficient cl or cs normal or is there any problem with my job that might lead to this rank deficiency?

As for the numerical instability itself, I was able to fix it by revising the next few lines as (see comments)
```
        # new code
        if numpy.max(numpy.abs(x)) > 1e20:
            raise scipy.linalg.LinAlgError
    except scipy.linalg.LinAlgError:
        d, t = numpy.linalg.eigh(m)
        idx = d>LINEAR_DEP_THRESHOLD
        t = t[:,idx] / numpy.sqrt(d[idx])
        tht = reduce(numpy.dot, (t.T.conj(), h, t))
        e, a = numpy.linalg.eigh(tht)
        a = numpy.dot(t, a)
        idx = e > -c**2
        cl = a[:nao,idx]
        cs = a[nao:,idx]
        # X = B A^{-1} = B A^T S (old equation)
        # x = cs.dot(cl.conj().T).dot(m) (old code, here dot(m) is a bug. The S matrix in B A^T S is neither m nor s)
        # new code below. Equivalent to x = cs.dot(numpy.linalg.inv(cl)) if cl is square and full-rank
        u, s, vt = numpy.linalg.svd(cl, full_matrices=False)
        idx = s>LINEAR_DEP_THRESHOLD
        cl_inv = vt[idx].conj().T / s[idx] @ u[:, idx].conj().T
        x = cs.dot(cl_inv)
    return x
```
This gave me x that did not diverge (norm of x is about 18). (I can create a pull request if this fix is correct.)

However, after this fix, my calculation still seemed problematic. There was a significant decrease in energy (mostly in E1) to some unreasonable value at the first cycle. Here are the total energy of first few iterations
init E= -17021.7296433587
cycle= 1 E= -38914.6403139796  delta_E= -2.19e+04  |g|= 66.7  |ddm|= 22.2
cycle= 2 E= -38925.3858762173  delta_E= -10.7  |g|= 14.7  |ddm|= 19.7
cycle= 3 E= -38859.0670998692  delta_E= 66.3  |g|= 22.5  |ddm|= 9.97
cycle= 4 E= -38954.7320424528  delta_E= -95.7  |g|= 2.13  |ddm|= 8.73
...
and the E1 and E_coul of the same iterations:
E1 = -21226.4302029105  E_coul = 4422.708655362758
E1 = -42762.61865404233  E_coul = 4065.9864358736986
E1 = -43879.92840974802  E_coul = 5172.550629341682
E1 = -42315.25614001084  E_coul = 3674.1971359525382
E1 = -43487.89319770682  E_coul = 4751.1692510649245
...
For comparison, the total energy without x2c is about -11623 Ha.

Any help or comments on what might go wrong would be greatly appreciated! I have tried other systems in other basis. Dy and Er in sarc-dkh2 basis both had the same problem; however, Dy and Er in cc-pvtz-dk basis were both fine; W in sarc-dkh2 was also fine. 

Tested with PySCF version 2.5.0.

**Comment by xubwa** [comment]
It's a basis set issue. X2C transformation is carried out under the uncontracted basis, and some basis such as segmented contracted basis set might have severe linear dependency problem in the basis's primitive functions. As you have observed, cc-pvtz-dk works fine, that's because correlation consistent set is a general contraction basis set and doesn't have the problem. There shouldn't be a fix for this problem, since in my personal view, these basis sets just should not be used for x2c calculation. Maybe stick with what cc sets have or try using the ano set if one want to use contracted basis set for x2c. 
Note, using such basis still have minor problems since they are contracted under different relativistic Hamiltonian. There are some newer cc sets contracted for sfx2c out there, the ano set also has a newer version for x2c. If one can afford to use uncontracted basis set, then dyall set can also be an option.
P.S., a 'wrong' fix for the problem can be doing the x2c transformation under the contracted basis set and pyscf actually has a keyword to do that. The xuncontract parameter can be manually set to False after creating the x2c object. Then in principle your calculation should be able to run and yield reasonable energy. But it's strongly not recommended. Solving the dirac equation under the contracted basis cannot guarantee the kinetic balance condition.

**Comment by xubwa** [comment]
It's a basis set issue. X2C transformation is carried out under the uncontracted basis, and some basis such as segmented contracted basis set might have severe linear dependency problem in the basis's primitive functions. As you have observed, cc-pvtz-dk works fine, that's because correlation consistent set is a general contraction basis set and don't have the problem. There shouldn't be a fix for this problem, since in my personal view, these basis sets just should not be used for x2c calculation. Maybe stick with what cc sets have or try using the ano set if one want to use contracted basis set for x2c. Note, using such basis still have problem since they are contracted under different relativistic Hamiltonian, there are some newer cc sets contractd for sfx2c out there, the ano set also has a newer version for x2c. If one can afford to use uncontracted basis set, then dyall set can also be a option.
P.S., a 'wrong' fix for the problem can be doing the x2c transformation under the contracted basis set and pyscf actually has a keyword to do that. The xuncontract parameter can be manually set to False after creating the x2c object. Then in principle your calculation should be able to run and yield reasonable energy. But it's strongly not recommended. Solving the dirac equation under the contracted basis cannot guarantee the kinetic balance condition.

**Comment by penglinq** [comment]
Thank you very much for the fast reply and for explaining and advising on the basis sets, Xubo! 

I tried either using the newer cc basis contracted for sfx2c you suggested ([cc-pwCVnZ-X2C](https://pubs.aip.org/aip/jcp/article/147/8/084108/195973), right?) or simply adding mf.with_x2c.xuncontract = False for the sarc-dkh2 basis, both solved my rank deficiency / divergence problem.

**Comment by penglinq** [comment]
I have two follow-up questions:
1. I checked the condition number of the uncontracted sarc-dkh2 basis for Dy at the gamma point - it's about 1e4, which is not very large. Here's what I looked at: 
```
xcell, _ = mf.with_x2c.get_xmol(cell)
S = xcell.pbc_intor('int1e_ovlp', hermi=1, kpts=[[0,0,0]])[0]
```
The largest eigenvalue of S is 5.6 and the smallest is 5.8e-4. To help understand why the uncontracted sarc-dkh2 basis leads to the numerical issues, is there any other quantity that I should look at or is there another reason?
2. Using a contracted basis (setting xuncontract=False) for x2c seems to yield much more negative energy. Is that normal? For example, I performed molecular GHF with x2c on the [DyCl6]3- octahedral complex. When I evaluated the energy of the converged GHF solution,  I got -14938.36 Ha with contracted basis and -14917.50 Ha with uncontracted basis.  Similarly for UHF with sfx2c, I got -14933.29 Ha with contract basis and -14914.91 Ha with uncontracted basis.

Thanks again.

**Comment by xubwa** [comment]
I checked the sarc-dkh basis, it looks fine, it's not that sort of problematic basis set, so I would expect everything runs fine. I tried it in molecular case, it's also fine. I'm not familiar with the periodic part of pyscf, so I have little idea on what's wrong here.
For your question on 'more negative energy', as I said, using xuncontract=False would only help you run through the calculation, but it is wrong, and the direction can be arbitrary. Unlike in nonrelativistic theory, the relativistic energy is not 'variational, with kinetic balance imposed in an improper way due to the contracted basis, the direction of energy change is indeed unpredictable.

**Comment by penglinq** [comment]
Indeed, as you pointed out, for all lanthanide molecular complexes (> 10) I tried, x2c worked with no issue. I guess I will look into the x2c with PBC code more carefully. Thanks for explaining more about the energy!

**Comment by sunqm** [comment]
By default, the X matrix in PBC-X2C is solved with the problematic Coulomb potential integrals. The V and W matrices in PBC-X2C are computed with the zero-net charge assumption to remove the singularity in these integrals. This is equivalent to add an infinity large constant to the integrals to cancel out the divergent part:
```math
\begin{pmatrix}
V + C1 & T \\
T & W/4c^2 + C2
\end{pmatrix}
```
The constants C1 and C2 are infinitely large and they can lead to arbitrary X matrix.

I'd recommend to using the atomix X approximation, by setting the `mf.with_df = 'ATOM1E'` . This should be more physical X matrix. See also the discussions in #2174
